### ISSUE TITLE
provide a peeking API to std.io.BufferedReader

### ISSUE THREAD

**Issue by andrewrk** [ziglang/zig#4501]
The point of BufferedInStream is to call `read()` on the underlying stream in bursts, rather than separate calls. E.g. to make 1 syscall to `read()` reading 999 bytes, rather than 999 syscalls to `read()` reading 1 byte each.

The point of PeekStream is to be able to "put back" bytes, so that the next call to `read()` will get them again. In order to accomplish this, PeekStream needs a buffer. That makes PeekStream also a BufferedInStream.

All we need to unify these two APIs is for BufferedInStream to have a parameter called `reserved_buffer_space` or something like this, and then it's guaranteed to be able to "put back" this many elements.

Related:
* #19298

**Comment by pierrec** [comment]
Hello,

Since I make use of io.PeekStream and it just got removed, I would like to contribute to this issue.
The idea is to simply implement the old methods on BufferedReader.

```
pub fn putBackByte(self: *Self, byte: u8) !void { ... }

pub fn putBackByte(self: *Self, byte: u8) !void { ... }
```

Would this be acceptable?

**Comment by nektro** [comment]
that seems backwards. there should be a `peek()` function that returns `br.buf[br.start]` and `read()` should eat the byte

**Comment by travisstaloch** [comment]
> that seems backwards. there should be a peek()

I agree that it would be nice to have a `peek()`.  How about `peek([]u8) !void` and `peekByte() !u8`?  

But previously `io.PeekStream()` had `putBackByte(u8)` and `putBack([]const u8)`.  So maybe we could have both `peek()` and `putBack()`?

**Comment by septemhill** [comment]
I have a question, how do we handle the data after `peek` ? use `read` to consume or add one more method `discard` to ignore the data (byte or slice) ?

**Comment by andrewrk** [comment]
Solved by #24329
